File: src/logic.go
package main

func sendRowToChannel(world []byte, row int, width int, dest chan<- byte) {
	for i := 0; i < width; i++ {
		dest <- world[(row*width)+i]
	}
}

// Calculate how many rows each thread will be responsible for
func rowsForEachThread(threads int, rows int) []int {
	s := make([]int, threads)
	for row := 0; row < rows; row++ {
		s[row%threads]++
	}
	return s
}

func customMod(index int, max int) int {
	if index >= max {
		return index - max
	} else if index < 0 {
		return index + max
	} else {
		return index
	}
}

// Rows is the number of rows to process
// Width is how wide each row is
// The offset is where the starting row is in the matrix
// The result will have size rows * width
func gameOfLifeTurn(matrix []byte, rows int, width int, offset int) []byte {
	mHeight := len(matrix) / width
	result := make([]byte, rows*width)
	for y := 0; y < rows; y++ {
		realY := y + offset
		for x := 0; x < width; x++ {

			result[(y*width)+x] = matrix[(realY*width)+x]

			var neighboursAlive = 0
			//Count alive cells in 3x3 grid
			for i := y - 1; i <= y+1; i++ {
				realI := i + offset
				for j := x - 1; j <= x+1; j++ {
					if matrix[(customMod(realI, mHeight)*width)+customMod(j, width)] == ALIVE {
						neighboursAlive++
					}
				}
			}
			if result[(y*width)+x] == ALIVE {
				neighboursAlive--
				if neighboursAlive == 2 || neighboursAlive == 3 {
					result[(y*width)+x] = ALIVE
				} else {
					result[(y*width)+x] = DEAD
				}
			} else if neighboursAlive == 3 {
				result[(y*width)+x] = ALIVE
			}
		}
	}
	return result
}

File: rust/src/logic.rs
use crate::World;

const ALIVE: u8 = 0xFF;
const DEAD: u8 = 0x00;

#[inline(always)]
fn custom_mod(index: i32, max: i32) -> i32 {
    if index >= max {
        return index - max
    } else if index < 0 {
        return index + max
    } else {
        return index
    }
}

#[inline(always)]
fn index(row: i32, column: i32, width: i32) -> usize {
    ((row * width) + column) as usize
}

pub fn calculate_row(world: &World, row: i32) -> Vec<u8> {

    let start = (world.width * row) as usize;
    let end = start + world.width as usize;
    let mut result: Vec<u8> = world.cells[start..end].iter().cloned().collect();

    for col  in 0..world.width  {
        let mut neighbours_alive = 0;

        for i in (row-1)..(row+2) {
            for j in (col-1)..(col+2) {
                let idx = index(custom_mod(i, world.height),custom_mod(j, world.width), world.width);
                if world.cells[idx] == ALIVE {
                    neighbours_alive = neighbours_alive + 1;
                }
            }
        }

        if world.cells[index(row, col, world.width)] == ALIVE {
            neighbours_alive = neighbours_alive - 1;
            if neighbours_alive == 2 || neighbours_alive == 3 {
                result[col as usize] = ALIVE
            } else {
                result[col as usize] = DEAD
            }
        } else if neighbours_alive == 3 {
            result[col as usize] = ALIVE
        }
    }

    return result
}
